### PR TITLE
various smaller logind fixes

### PR DESCRIPTION
**Pull request by poettering**
Mostly about fixing #8035, but also other stuff encountered while working on this. Mostly minor fixes.

### CODE REVIEW

**Comment by sourcejedi** [comment]
Reproducing #8035 seems really annoying, a 10 minute timejump did not triggering all the 3 minute watchdog timeouts, only one of them and it was not logind.  For future testing, I've lowered the logind watchdog timeout to 15 seconds.

Doing so happened to trigger a watchdog timeout, and then minutes later (on VT switch?) a very similar looking issue with a DRM fd.  It feels to me like this shows a common root cause which we should address as a whole, I could be wrong though.

DRM fd assert backtrace: https://github.com/systemd/systemd/issues/8035#issuecomment-368670186

File: src/login/logind-session-device.c [hunk 1]
```diff
@@ -107,17 +107,15 @@ static int session_device_notify(SessionDevice *sd, enum SessionDeviceNotificati
 }
 
 static int sd_eviocrevoke(int fd) {
-        static bool warned;
-        int r;
+        static bool warned = false;
```
Comment (keszybz) [inline src/login/logind-session-device.c:110]: Patch comment suggests that `warned` wasn't initialized properly, but that's not the case. It was initialized implicitly. So the rework is fine, but the commit message is misleading ;)

File: src/login/logind-session-device.c [hunk 2]
```diff
@@ -107,17 +107,15 @@ static int session_device_notify(SessionDevice *sd, enum SessionDeviceNotificati
 }
 
 static int sd_eviocrevoke(int fd) {
```
Comment (keszybz) [inline src/login/logind-session-device.c:109]: Maybe make this return void?

File: src/login/logind.c [hunk 3]
```diff
@@ -453,6 +453,12 @@ static int manager_attach_fds(Manager *m) {
                         continue;
                 }
 
+                if (!S_ISCHR(st.st_mode) && !S_ISBLK(st.st_mode)) {
+                        log_debug("Device fd doesn't actually point to device node: %m");
```
Comment (keszybz) [inline src/login/logind.c:457]: to *a* device node? Maybe include fd value in the message?

File: src/login/logind-session-device.c [hunk 4]
```diff
@@ -420,7 +420,7 @@ void session_device_free(SessionDevice *sd) {
 
         session_device_stop(sd);
         session_device_notify(sd, SESSION_DEVICE_RELEASE);
-        close_nointr(sd->fd);
+        safe_close(sd->fd);
```
Comment (keszybz) [inline src/login/logind-session-device.c:423]: Oooh, nice bug ;)

**Comment by keszybz** [comment]
On second thought, none of those suggestions are important enough to delay this, and can just as well be applied as separate patches, so let's merge.

**Comment by sourcejedi** [comment]
I guess so.  Obviously the removed assert does not trigger anymore.

Suspending the host last night (and resuming this morning) only caused a watchdog restart.  However, I had a look whether that the "spice vdagent tablet" input device was removed and re-added, as the crash happened when handling the udev remove event.  Unlike the original report, `journalctl` did not show a message after the suspend looking like

    kernel: input: spice vdagent tablet as /devices/virtual/input/input6

So the answer to "can I reproduce the scenario which caused the same assert" is not so easy :(.  Perhaps there are a few more variable factors in my test.

**Pull request merged**